packages/cli/src/lib/features.ts

```diff
@@ -0,0 +1,163 @@
+  'backend-plugin-module',
+  'frontend-plugin',
+  'frontend-plugin-module',
+];
+
+// A list of the feature types we want to extract from the project
+// and include in the metadata
+const targetFeatureTypes: BackstagePackageFeatureType[] = [
+  '@backstage/BackendFeature',
+  '@backstage/BackstagePlugin',
+  '@backstage/FrontendPlugin',
+  '@backstage/FrontendModule',
+];
+
+// Returns all valid Backstage package features from a project
+// for all entry points.
+export function getFeaturesMetadata(
+  project: Project,
+  role: PackageRole,
+  dir: string,
+  entryPoints: EntryPoint[],
+): BackstagePackageFeature[] {
+  if (!targetPackageRoles.includes(role)) {
+    return [];
+  }
+
+  return entryPoints
+    .flatMap(entryPoint => getEntryPointExports(entryPoint, project, dir))
+    .filter(isDefaultExport)
+    .filter(isTargetFeatureType)
+    .map(toBackstagePackageFeature);
+}
+
+type EntryPointExport = {
+  name: string;
+  location: string;
+  type: BackstagePackageFeatureType;
+};
+
+// Returns all exports (default and named) from an entry point
+// that are valid Backstage package features
+export function getEntryPointExports(
+  { mount: location, path, ext }: EntryPoint,
+  project: Project,
+  dir: string,
+): EntryPointExport[] {
+  const fullFilePath = resolvePath(dir, path);
+  const sourceFile = project.getSourceFile(fullFilePath);
+
+  if (!sourceFile || (ext !== '.ts' && ext !== '.tsx')) {
+    return [];
+  }
+
+  const exports = [];
+
+  for (const exportSymbol of sourceFile.getExportSymbols()) {
+    const declaration = exportSymbol.getDeclarations()[0];
+    const exportName = declaration.getSymbol()?.getName();
+    let exportType: Type<ts.Type> | undefined;
+
+    if (declaration) {
+      if (declaration.isKind(SyntaxKind.ExportAssignment)) {
+        exportType = declaration.getExpression().getType();
+      } else if (declaration.isKind(SyntaxKind.ExportSpecifier)) {
+        if (!declaration.isTypeOnly()) {
+          exportType = declaration.getType();
+        }
+      } else if (declaration.isKind(SyntaxKind.VariableDeclaration)) {
+        exportType = declaration.getType();
+      }
+    }
+
+    if (exportName && exportType) {
+      const $$type = getBackstagePackageFeature$$TypeFromType(exportType);
+
+      if ($$type) {
+        exports.push({
+          name: exportName,
+          location,
+          type: $$type,
+        });
+      }
+    }
+  }
+
+  return exports;
+}
+
+// Given a TS type, returns the Backstage package feature $$type value
+function getBackstagePackageFeature$$TypeFromType(
+  type: Type,
+): BackstagePackageFeatureType | null {
+  // Returns the concrete type of a generic type
+  const exportType = type.getTargetType() ?? type;
+
+  for (const property of exportType.getProperties()) {
+    if (property.getName() === '$$type') {
+      const $$type = property
+        .getValueDeclaration()
+        ?.getText()
+        .match(/(?<type>@backstage\/\w+)/)?.groups?.type;
```

Comment:
Really no need to filter out just the `@backstage/*` types here right? We have the specific list of types we support anyway

Reply:
This is needed to pull out just the @backstage/* portion as the entire declaration can be like `readonly $$type: '@backstage/*';`